### PR TITLE
CMake rework - modernization and simplification

### PR DESCRIPTION
**Pull request by kreczko**
@tunnell @jpbrodsky @jecutter 
This arose from a private discussion with @mszydagis on LZ gitlab.

## Summary
While looking at the 2.1.0beta I realised that there is a difference in how headers are stored in the repo and how they are installed. The repository structure is closer to current standards that scope headers (e.g. `#include "NEST/NEST.hh"` vs `#include "NEST.hh"`.

I took this opportunity to fix a few things in the current CMake as well as to modernize (e.g. `NEST::Core` target) and simplify (current CMake is a mix of old and new language styles).

With a simple
```cmake
target_link_libraries(<your target> PUBLIC NEST::Core)
```
all necessary include paths are imported and allow for both nested and direct header inclusions:
```c++
#include "NEST.hh"
#include "NEST/NEST.hh"

#include "VDetector.hh"
#include "Detectors/VDetector.hh"

#include "NEST/G4/NESTStackingAction.hh"
#include "NESTStackingAction.hh"
```

Ideally, `Detectors` would also go under `NEST/`, but that would mismatch the repository structure.
Full install structure is shown below.

The old CMake produced both STATIC (only used for local executables) and SHARED versions of the core library, the current produces STATIC by default and SHARED with [BUILD_SHARED_LIBS=ON](https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html). If the old behaviour is required, I can add an object library as an intermediate step.

Feedback is welcome.

## Current 2.1.0beta install structure (CentOS7)
```bash
 tree -L 2 install/
install/
├── NESTConfig-noconfig.cmake
├── NESTConfig.cmake
├── NESTConfigVersion.cmake
├── bareNEST
├── include
│   ├── DetectorExample_XENON10.hh
│   ├── LUX_Run03.hh
│   ├── NEST.hh
│   ├── RandomGen.hh
│   ├── TestSpectra.hh
│   ├── VDetector.hh
│   ├── Xed_CWRU_Dahl.hh
│   ├── Z3_FSR_HA.hh
│   ├── analysis.hh
│   ├── testNEST.hh
│   └── xoroshiro.hh
├── lib
│   └── libNEST.so
└── testNEST
```

## Install structure with this MR (G4=ON, BUILD_ROOT=ON)

```bash
```shell
 tree -L 4 install/
install/
├── bin
│   ├── bareNEST
│   ├── rootNEST
│   └── testNEST
├── include
│   ├── Detectors
│   │   ├── DetectorExample_XENON10.hh
│   │   ├── LUX_Run03.hh
│   │   ├── VDetector.hh
│   │   ├── Xed_CWRU_Dahl.hh
│   │   └── Z3_FSR_HA.hh
│   └── NEST
│       ├── G4
│       │   ├── NESTProc.hh
│       │   └── NESTStackingAction.hh
│       ├── NEST.hh
│       ├── RandomGen.hh
│       ├── TestSpectra.hh
│       ├── analysis.hh
│       ├── testNEST.hh
│       └── xoroshiro.hh
└── lib64
    ├── cmake
    │   └── NEST
    │       ├── NESTConfig.cmake
    │       ├── NESTConfigVersion.cmake
    │       ├── NESTTargets-noconfig.cmake
    │       └── NESTTargets.cmake
    ├── libCore.a
    └── libNESTG4.a
```

### CODE REVIEW

File: G4integration/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+add_library(
+    NESTG4
+    ${NEST_G4_SOURCES}
+    ${NEST_G4_HEADERS}
+)
+add_library(NEST::NESTG4 ALIAS NESTG4)
+
+target_compile_features(NESTG4 PUBLIC cxx_std_17)
```
Comment (jpbrodsky) [inline G4integration/CMakeLists.txt:20]: Again, does this need to be 17?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,119 +1,116 @@
...
+  Core PUBLIC
+  $<INSTALL_INTERFACE:include>
+  $<INSTALL_INTERFACE:include/NEST>
+  $<INSTALL_INTERFACE:include/Detectors>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/NEST>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/Detectors>
+)
+
+target_compile_features(Core PUBLIC cxx_std_17)
```
[the remark this answers is not in the document]
Comment (jpbrodsky) [reply CMakeLists.txt:43]: (Notably, ROOT is often compiled with 14 or even 11 with non-GCC compilers, due to weirdness in how ROOT handles its compiler feature checks)

[the remark this answers is not in the document]
Comment (jpbrodsky) [reply CMakeLists.txt:43]: So I wouldn't force this into 17 if it's not necessary.

[the remark this answers is not in the document]
Comment (kreczko) [reply CMakeLists.txt:43]: it is not necessary - will remove.

File: G4integration/CMakeLists.txt (hunk 1 above)
Comment (kreczko) [reply G4integration/CMakeLists.txt]: Removed

**Comment by mszydagis** [comment]
One more thing I need your help with on this, @kreczko and @jpbrodsky  I have added a new executable (like rootNEST) in the examples sub-folder. But I have been manually, for weeks now, compiling it awkwardly just separately on my own. It is called PulseShape.cpp  I don't know enough about making makefiles (pun intended haha!) Could one of you please add this new pulse shape code to the cmake file, alongside rootNEST (and testNEST and bareNEST etc.) in same way? Thank you!! (And then I can tag the release tomorrow hopefully, woo-hoo!) This will also add the advantage of being able to call headers and such from everywhere. Thanks again

**Comment by kreczko** [comment]
@mszydagis 
This is now done in d347e57.

For future reference, this is the pattern for the example directory:
In `examples/CMakeLists.txt` add

```cmake
add_executable(<binary name> <sources>)
target_link_libraries(<binary name>PUBLIC NEST::Core)
list(APPEND EXAMPLES <binary name>)
```

BTW: PulseShape.cpp was not compiling (missing `#include <algorithm>` for `std::sort`) - I've rebased onto master and fixed the missing header.

**Comment by mszydagis** [comment]
thank you for this. And good catch on missing header: for some reason worked on my laptop.

**Pull request merged**